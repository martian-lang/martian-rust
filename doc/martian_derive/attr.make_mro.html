<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="When this attribute is applied to the `MartianMain` or `MartianStage` trait implementation of a stage struct, it derives the trait `MroMaker` to the stage struct, which lets you generate the mro corresponding to the stage."><title>make_mro in martian_derive - Rust</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceSerif4-Regular-46f98efaafac5295.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/FiraSans-Regular-018c141bf0843ffd.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/FiraSans-Medium-8f9a781e4970d388.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceSerif4-Bold-a2c9cd1067f8b328.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2"><link rel="stylesheet" href="../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../static.files/rustdoc-cb6f1f67f1bcd037.css" id="mainThemeStyle"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="martian_derive" data-themes="" data-resource-suffix="" data-rustdoc-version="1.73.0 (cc66ad468 2023-10-03)" data-channel="1.73.0" data-search-js="search-6dfdfced5eff6596.js" data-settings-js="settings-de11bff964e9d4e5.js" data-settings-css="settings-8c76f75bfb6bd192.css" data-theme-light-css="light-1596385f77d47ef2.css" data-theme-dark-css="dark-0a43001d3fc2282c.css" data-theme-ayu-css="ayu-fd19013d6ce078bf.css" ><script src="../static.files/storage-db41da1a38ea3cb8.js"></script><script defer src="sidebar-items.js"></script><script defer src="../static.files/main-0795b7d26be81095.js"></script><noscript><link rel="stylesheet" media="(prefers-color-scheme:light)" href="../static.files/light-1596385f77d47ef2.css"><link rel="stylesheet" media="(prefers-color-scheme:dark)" href="../static.files/dark-0a43001d3fc2282c.css"><link rel="stylesheet" href="../static.files/noscript-cffde32267a19fd6.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-16x16-8b506e7a72182f1c.png"><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc attr"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle">&#9776;</button><a class="logo-container" href="../martian_derive/index.html"><img class="rust-logo" src="../static.files/rust-logo-151179464ae7ed46.svg" alt="logo"></a><h2></h2></nav><nav class="sidebar"><a class="logo-container" href="../martian_derive/index.html"><img class="rust-logo" src="../static.files/rust-logo-151179464ae7ed46.svg" alt="logo"></a><div class="sidebar-elems"><h2><a href="index.html">In martian_derive</a></h2></div></nav><main><div class="width-limiter"><nav class="sub"><form class="search-form"><span></span><input class="search-input" name="search" aria-label="Run search in the documentation" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><div id="help-button" title="help" tabindex="-1"><a href="../help.html">?</a></div><div id="settings-menu" tabindex="-1"><a href="../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../static.files/wheel-7b819b6101059cd0.svg"></a></div></form></nav><section id="main-content" class="content"><div class="main-heading"><h1>Attribute Macro <a href="index.html">martian_derive</a>::<wbr><a class="attr" href="#">make_mro</a><button id="copy-path" title="Copy item path to clipboard"><img src="../static.files/clipboard-7571035ce49a181d.svg" width="19" height="18" alt="Copy item path"></button></h1><span class="out-of-band"><a class="src" href="../src/martian_derive/lib.rs.html#38-210">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><pre class="rust item-decl"><code>#[make_mro]</code></pre><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>When this attribute is applied to the <code>MartianMain</code> or <code>MartianStage</code> trait implementation of
a stage struct, it derives the trait <code>MroMaker</code> to the stage struct, which lets you generate
the mro corresponding to the stage.</p>
<p>You can optionally specify <code>mem_gb</code>, <code>threads</code>, <code>vmem_gb</code> and <code>volatile</code> within this proc-macro.
For example, use <code>#[make_mro(mem_gb = 4, threads = 2]</code> for setting <code>mem_gb</code> and <code>threads</code> that would
appear in the <code>using()</code> section of the mro definition.</p>
<p>You can also set the stage name here. By default, the stage name in the mro is the SHOUTY_SNAKE_CASE version
of the stage struct name. You can override that using: <code>#[make_mro(mem_gb = 2, stage_name = MY_CUSTOM_NAME)]</code></p>
<p>For examples on how to use it and customize, take a look at <code>tests/test_full_mro.rs</code></p>
</div></details></section></div></main></body></html>