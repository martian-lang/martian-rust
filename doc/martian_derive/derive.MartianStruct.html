<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Structs which are used as associated types in `MartianMain` or `MartianStage` traits need to implement `MartianStruct`. You can derive it using `#[derive(MartianStruct)]` Each field in the struct needs to implement `AsMartianBlanketType`. This is implemented for most basic Rust types. You can derive it on custom structs or enum using `#[derive(MartianType)]`."><meta name="keywords" content="rust, rustlang, rust-lang, MartianStruct"><title>MartianStruct in martian_derive - Rust</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceSerif4-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../FiraSans-Regular.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../FiraSans-Medium.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceCodePro-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceSerif4-Bold.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceCodePro-Semibold.ttf.woff2"><link rel="stylesheet" href="../normalize.css"><link rel="stylesheet" href="../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" href="../ayu.css" disabled><link rel="stylesheet" href="../dark.css" disabled><link rel="stylesheet" href="../light.css" id="themeStyle"><script id="default-settings" ></script><script src="../storage.js"></script><script defer src="sidebar-items.js"></script><script defer src="../main.js"></script><noscript><link rel="stylesheet" href="../noscript.css"></noscript><link rel="alternate icon" type="image/png" href="../favicon-16x16.png"><link rel="alternate icon" type="image/png" href="../favicon-32x32.png"><link rel="icon" type="image/svg+xml" href="../favicon.svg"></head><body class="rustdoc derive"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle">&#9776;</button><a class="sidebar-logo" href="../martian_derive/index.html"><div class="logo-container"><img class="rust-logo" src="../rust-logo.svg" alt="logo"></div></a><h2></h2></nav><nav class="sidebar"><a class="sidebar-logo" href="../martian_derive/index.html"><div class="logo-container"><img class="rust-logo" src="../rust-logo.svg" alt="logo"></div></a><div class="sidebar-elems"><h2><a href="index.html">In martian_derive</a></h2></div></nav><main><div class="width-limiter"><nav class="sub"><form class="search-form"><span></span><input class="search-input" name="search" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><div id="help-button" title="help" tabindex="-1"><a href="../help.html">?</a></div><div id="settings-menu" tabindex="-1"><a href="../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../wheel.svg"></a></div></form></nav><section id="main-content" class="content"><div class="main-heading"><h1 class="fqn">Derive Macro <a href="index.html">martian_derive</a>::<wbr><a class="derive" href="#">MartianStruct</a><button id="copy-path" onclick="copy_path(this)" title="Copy item path to clipboard"><img src="../clipboard.svg" width="19" height="18" alt="Copy item path"></button></h1><span class="out-of-band"><a class="srclink" href="../src/martian_derive/lib.rs.html#408-575">source</a> · <a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span></div><div class="item-decl"><pre class="rust derive"><code>#[derive(MartianStruct)]
{
    // Attributes available to this derive:
    #[mro_retain]
    #[mro_type]
}
</code></pre></div><details class="rustdoc-toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Structs which are used as associated types in <code>MartianMain</code> or <code>MartianStage</code>
traits need to implement <code>MartianStruct</code>. You can derive it using <code>#[derive(MartianStruct)]</code>
Each field in the struct needs to implement <code>AsMartianBlanketType</code>. This is implemented
for most basic Rust types. You can derive it on custom structs or enum using <code>#[derive(MartianType)]</code>.</p>
<p>You can optionally annotate the mro type manually for each field (<code>#[mro_type = &quot;map[]&quot;]</code> for example).
This is useful if the datatype for the field is from a third party crate.</p>
<p>You can optionally add a field to the “retain” section of the mro using <code>#[mro_retain]</code>.</p>
</div></details></section></div></main><div id="rustdoc-vars" data-root-path="../" data-current-crate="martian_derive" data-themes="ayu,dark,light" data-resource-suffix="" data-rustdoc-version="1.66.0 (69f9c33d7 2022-12-12)" ></div></body></html>