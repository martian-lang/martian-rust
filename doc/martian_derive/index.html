<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="API documentation for the Rust `martian_derive` crate."><meta name="keywords" content="rust, rustlang, rust-lang, martian_derive"><title>martian_derive - Rust</title><link rel="stylesheet" type="text/css" href="../normalize.css"><link rel="stylesheet" type="text/css" href="../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../light.css"  id="themeStyle"><link rel="stylesheet" type="text/css" href="../dark.css" disabled ><link rel="stylesheet" type="text/css" href="../ayu.css" disabled ><script id="default-settings"></script><script src="../storage.js"></script><script src="../crates.js"></script><noscript><link rel="stylesheet" href="../noscript.css"></noscript><link rel="icon" type="image/svg+xml" href="../favicon.svg">
<link rel="alternate icon" type="image/png" href="../favicon-16x16.png">
<link rel="alternate icon" type="image/png" href="../favicon-32x32.png"><style type="text/css">#crate-search{background-image:url("../down-arrow.svg");}</style></head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="sidebar"><div class="sidebar-menu" role="button">&#9776;</div><a href='../martian_derive/index.html'><div class='logo-container rust-logo'><img src='../rust-logo.png' alt='logo'></div></a><h2 class="location">Crate martian_derive</h2><div class="block version"><p>Version 0.24.0</p></div><div class="sidebar-elems"><a id="all-types" href="all.html"><p>See all martian_derive's items</p></a><div class="block items"><ul><li><a href="#macros">Macros</a></li></ul></div><div id="sidebar-vars" data-name="martian_derive" data-ty="mod" data-relpath=""></div><script defer src="sidebar-items.js"></script></div></nav><div class="theme-picker"><button id="theme-picker" aria-label="Pick another theme!" aria-haspopup="menu" title="themes"><img src="../brush.svg" width="18" height="18" alt="Pick another theme!"></button><div id="theme-choices" role="menu"></div></div><nav class="sub"><form class="search-form"><div class="search-container"><div><select id="crate-search"><option value="All crates">All crates</option></select><input class="search-input" name="search" disabled autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"></div><button type="button" id="help-button" title="help">?</button><a id="settings-menu" href="../settings.html" title="settings"><img src="../wheel.svg" width="18" height="18" alt="Change settings"></a></div></form></nav><section id="main" class="content"><h1 class="fqn"><span class="in-band">Crate <a class="mod" href="#">martian_derive</a><button id="copy-path" onclick="copy_path(this)" title="copy path"><img src="../clipboard.svg" width="19" height="18" alt="Copy item import" title="Copy item import to clipboard"></button></span><span class="out-of-band"><span id="render-detail"><a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span><a class="srclink" href="../src/martian_derive/lib.rs.html#1-890" title="goto source code">[src]</a></span></h1><h2 id="macros" class="section-header"><a href="#macros">Macros</a></h2>
<table><tr class="module-item"><td><a class="macro" href="macro.martian_filetype.html" title="martian_derive::martian_filetype macro">martian_filetype</a></td><td class="docblock-short"><p>A macro to define a new struct that implements <code>MartianFileType</code> trait</p>
</td></tr></table><h2 id="attributes" class="section-header"><a href="#attributes">Attribute Macros</a></h2>
<table><tr class="module-item"><td><a class="attr" href="attr.make_mro.html" title="martian_derive::make_mro attr">make_mro</a></td><td class="docblock-short"><p>When this attribute is applied to the <code>MartianMain</code> or <code>MartianStage</code> trait implementation of
a stage struct, it derives the trait <code>MroMaker</code> to the stage struct, which lets you generate
the mro corresponding to the stage.</p>
</td></tr></table><h2 id="derives" class="section-header"><a href="#derives">Derive Macros</a></h2>
<table><tr class="module-item"><td><a class="derive" href="derive.MartianStruct.html" title="martian_derive::MartianStruct derive">MartianStruct</a></td><td class="docblock-short"><p>Structs which are used as associated types in <code>MartianMain</code> or <code>MartianStage</code>
traits need to implement <code>MartianStruct</code>. You can derive it using <code>#[derive(MartianStruct)]</code>
Each field in the struct needs to implement <code>AsMartianBlanketType</code>. This is implemented
for most basic Rust types. You can derive it on custom structs or enum using <code>#[derive(MartianType)]</code>.</p>
</td></tr><tr class="module-item"><td><a class="derive" href="derive.MartianType.html" title="martian_derive::MartianType derive">MartianType</a></td><td class="docblock-short"><p>Custom types which are fields of a <code>MartianStruct</code> need to implement <code>AsMartianBlanketType</code>.
You can derive that trait on an enum or struct using <code>#[derive(MartianType)]</code></p>
</td></tr></table></section><section id="search" class="content hidden"></section><div id="rustdoc-vars" data-root-path="../" data-current-crate="martian_derive" data-search-index-js="../search-index.js" data-search-js="../search.js"></div><script src="../main.js"></script></body></html>