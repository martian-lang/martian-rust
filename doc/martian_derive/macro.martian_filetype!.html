<!DOCTYPE html>
<html lang="en">
<head>
    <meta http-equiv="refresh" content="0;URL=macro.martian_filetype.html">
    <title>Redirection</title>
</head>
<body>
    <p>Redirecting to <a href="macro.martian_filetype.html">macro.martian_filetype.html</a>...</p>
    <script>location.replace("macro.martian_filetype.html" + location.search + location.hash);</script>
</body>
</html>