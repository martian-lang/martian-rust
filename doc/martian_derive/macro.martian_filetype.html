<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="A macro to define a new struct that implements `MartianFileType` trait"><meta name="keywords" content="rust, rustlang, rust-lang, martian_filetype"><title>martian_filetype in martian_derive - Rust</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceSerif4-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../FiraSans-Regular.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../FiraSans-Medium.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceCodePro-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceSerif4-Bold.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceCodePro-Semibold.ttf.woff2"><link rel="stylesheet" href="../normalize.css"><link rel="stylesheet" href="../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" href="../ayu.css" disabled><link rel="stylesheet" href="../dark.css" disabled><link rel="stylesheet" href="../light.css" id="themeStyle"><script id="default-settings" ></script><script src="../storage.js"></script><script defer src="sidebar-items.js"></script><script defer src="../main.js"></script><noscript><link rel="stylesheet" href="../noscript.css"></noscript><link rel="alternate icon" type="image/png" href="../favicon-16x16.png"><link rel="alternate icon" type="image/png" href="../favicon-32x32.png"><link rel="icon" type="image/svg+xml" href="../favicon.svg"></head><body class="rustdoc macro"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle">&#9776;</button><a class="sidebar-logo" href="../martian_derive/index.html"><div class="logo-container"><img class="rust-logo" src="../rust-logo.svg" alt="logo"></div></a><h2></h2></nav><nav class="sidebar"><a class="sidebar-logo" href="../martian_derive/index.html"><div class="logo-container"><img class="rust-logo" src="../rust-logo.svg" alt="logo"></div></a><div class="sidebar-elems"><h2><a href="index.html">In martian_derive</a></h2></div></nav><main><div class="width-limiter"><nav class="sub"><form class="search-form"><span></span><input class="search-input" name="search" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><div id="help-button" title="help" tabindex="-1"><a href="../help.html">?</a></div><div id="settings-menu" tabindex="-1"><a href="../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../wheel.svg"></a></div></form></nav><section id="main-content" class="content"><div class="main-heading"><h1 class="fqn">Macro <a href="index.html">martian_derive</a>::<wbr><a class="macro" href="#">martian_filetype</a><button id="copy-path" onclick="copy_path(this)" title="Copy item path to clipboard"><img src="../clipboard.svg" width="19" height="18" alt="Copy item path"></button></h1><span class="out-of-band"><a class="srclink" href="../src/martian_derive/lib.rs.html#709-892">source</a> · <a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span></div><div class="item-decl"><pre class="rust macro"><code>martian_filetype!() { /* proc-macro */ }</code></pre></div><details class="rustdoc-toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>A macro to define a new struct that implements <code>MartianFileType</code> trait</p>
<p>Because this is a procedural macro, as of now, you can only define it
outside functions because they cannot be expanded to statements.</p>

<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="kw">use </span>serde::{Serialize, Deserialize};
<span class="kw">use </span>martian_derive::martian_filetype;
<span class="kw">use </span>martian::MartianFileType;
<span class="macro">martian_filetype! </span>{ TxtFile, <span class="string">&quot;txt&quot; </span>}
<span class="macro">martian_filetype! </span>{ BamIndexFile, <span class="string">&quot;bam.bai&quot; </span>}

<span class="macro">assert_eq!</span>(TxtFile::extension(), <span class="string">&quot;txt&quot;</span>);
<span class="macro">assert_eq!</span>(
    BamIndexFile::new(<span class="string">&quot;/path/to/folder&quot;</span>, <span class="string">&quot;filename&quot;</span>).as_ref(),
    std::path::Path::new(<span class="string">&quot;/path/to/folder/filename.bam.bai&quot;</span>)
)</code></pre></div>
</div></details></section></div></main><div id="rustdoc-vars" data-root-path="../" data-current-crate="martian_derive" data-themes="ayu,dark,light" data-resource-suffix="" data-rustdoc-version="1.66.0 (69f9c33d7 2022-12-12)" ></div></body></html>