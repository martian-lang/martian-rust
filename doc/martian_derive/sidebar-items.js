window.SIDEBAR_ITEMS = {"attr":["make_mro"],"derive":["MartianStruct","MartianType"],"macro":["martian_filetype"]};