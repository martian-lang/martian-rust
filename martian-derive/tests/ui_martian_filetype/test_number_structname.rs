
use martian_derive::martian_filetype;

martian_filetype! {4File, txt}

fn main() {}