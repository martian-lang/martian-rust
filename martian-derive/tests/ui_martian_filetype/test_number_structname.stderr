error: The first item `4File` in the martian_filetype! macro should be a valid identifier that can be used as a struct name.
 --> $DIR/test_number_structname.rs:4:20
  |
4 | martian_filetype! {4File, txt}
  |                    ^^^^^^^^^^
